### PR TITLE
Add support for UTF8 clipboard

### PR DESCRIPTION
**Pull request by pdlan**

### CODE REVIEW

**Comment by antenore** [comment]
WOW!

+1 for this PR

File: libvncserver/rfbserver.c [hunk 1]
```diff
@@ -596,6 +604,10 @@ rfbClientConnectionGone(rfbClientPtr cl)
 	deflateEnd( &(cl->compStream) );
     }
 
+    if (cl->extClipboardData != NULL) {
```
Comment (bk138) [inline libvncserver/rfbserver.c:607]: free(NULL) is okay, you can save 2 LOC here :-)

File: libvncserver/rfbserver.c [hunk 2]
```diff
@@ -2013,6 +2028,170 @@ rfbBool rfbProcessFileTransfer(rfbClientPtr cl, uint8_t contentType, uint8_t con
...
+        0xFF, 0xFF, 0xFF, 0xF8, /* -8 */
+        0x17, 0x00, 0x00, 0x01, /* text, request, peek, provide */
+        0x00, 0x10, 0x00, 0x00, /* max size is 1MiB */
+    };
+    if (rfbWriteExact(cl, buf, sizeof(buf)) < 0) {
+        rfbLogPerror("rfbSendExtendedClipboardCapability: write");
+        rfbCloseClient(cl);
+        return FALSE;
+    }
+    rfbStatRecordMessageSent(cl, rfbServerCutText, 16, 16);
```
Comment (bk138) [inline libvncserver/rfbserver.c:2045]: maybe `sizeof(buf)` here as well? but a really minor comment from my side...

File: libvncserver/rfbserver.c [hunk 3]
```diff
@@ -2013,6 +2028,170 @@ rfbBool rfbProcessFileTransfer(rfbClientPtr cl, uint8_t contentType, uint8_t con
...
+        0x03, 0x00, 0x00, 0x00,
+        0xFF, 0xFF, 0xFF, 0xFC, /* -4 */
+        0x08, 0x00, 0x00, 0x01, /* only text */
+    };
+    if (rfbWriteExact(cl, buf, sizeof(buf)) < 0) {
+        rfbLogPerror("rfbSendExtendedClipboardNotify: write");
+        rfbCloseClient(cl);
+        return FALSE;
+    }
+    rfbStatRecordMessageSent(cl, rfbServerCutText, 12, 12);
```
Comment (bk138) [inline libvncserver/rfbserver.c:2061]: Me thinks 16 bytes were written?

File: libvncserver/rfbserver.c [hunk 4]
```diff
@@ -2013,6 +2028,170 @@ rfbBool rfbProcessFileTransfer(rfbClientPtr cl, uint8_t contentType, uint8_t con
...
+            return FALSE;
+        }
+        if (i == 0) {
+            /* text */
+            if (!cl->viewOnly) {
+                cl->screen->setXCutTextUTF8(buf, size, cl);
+            }
+        }
+    }
+    if (buf != NULL) {
```
Comment (bk138) [inline libvncserver/rfbserver.c:2187]: Same here, free(NULL) is ok.

File: libvncserver/rfbserver.c [hunk 5]
```diff
@@ -2013,6 +2028,170 @@ rfbBool rfbProcessFileTransfer(rfbClientPtr cl, uint8_t contentType, uint8_t con
...
+            rfbLogPerror("rfbProcessExtendedServerCutTextData: failed to allocate memory");
+            inflateEnd(&stream);
+            rfbCloseClient(cl);
+            return FALSE;
+        }
+        stream.avail_out = size;
+        stream.next_out = buf;
+        if (inflate(&stream, Z_NO_FLUSH) != Z_OK) {
+            rfbLogPerror("rfbProcessExtendedServerCutTextData: zlib inflation error");
+            if (buf != NULL) {
```
Comment (bk138) [inline libvncserver/rfbserver.c:2173]: free(buf) is sufficient

File: libvncserver/rfbserver.c [hunk 6]
```diff
@@ -2621,6 +2818,13 @@ rfbProcessClientNormalMessage(rfbClientPtr cl)
 
 	msg.cct.length = Swap32IfLE(msg.cct.length);
 
+#ifdef LIBVNCSERVER_HAVE_LIBZ
+    if (cl->enableExtendedClipboard && (msg.cct.length & 0x80000000)) {
```
Comment (bk138) [inline libvncserver/rfbserver.c:2822]: Can you add an explaining comment here?

**Review by bk138 — CHANGES_REQUESTED** [review]
This is a very finely crafted PR, hats off! 

I have some very minor change requests/comments regarding the code.

I did _not_ check the logic against the spec. Can you provide some details on where/how you tested this in the commit message?

File: libvncserver/rfbserver.c (hunk 3 above)
Comment (pdlan) [reply libvncserver/rfbserver.c:2061]: Sorry it's my typo. The buffer size should be 12 but it seems that extra four zero bytes happen to be a empty FramebufferUpdate so the stream will not be corrupted. I will fix it.

**Comment by antenore** [comment]
Isn't this implemented through https://github.com/LibVNC/libvncserver/pull/468 ?

**Comment by Neustradamus** [comment]
Solves:
- https://github.com/LibVNC/libvncserver/issues/399

**Pull request merged**